models/flask_service.py

```diff
@@ -0,0 +1,127 @@
+
+app = Flask(__name__)
+api = Api(app)
+
+
+def to_dense(x):
+    return np.asarray(x.todense())
+
+
+def load_resources():
+    with open("users/preprocessing_pipeline.pkl", "rb") as f:
+        pipeline = pickle.load(f)
+    loaded_model = tf.keras.models.load_model("users/keras_model.keras")
+    return pipeline, loaded_model
+
+
+preprocessing_pipeline, model = load_resources()
+
+
+class LengthTransformer(BaseEstimator, TransformerMixin):
+    def __init__(self, columns):
+        self.columns = columns
+
+    def fit(self, X, y=None):
+        return self
+
+    def transform(self, X):
+        X_copy = X.copy()
+        for col in self.columns:
+            X_copy[col + "_length"] = X_copy[col].apply(
+                lambda x: len(str(x)) if pd.notnull(x) else 0
+            )
+        return X_copy
+
+
+class IsNullTransformer(BaseEstimator, TransformerMixin):
+    def __init__(self, columns):
+        self.columns = columns
+
+    def fit(self, X, y=None):
+        return self
+
+    def transform(self, X):
+        X_copy = X.copy()
+        for col in self.columns:
+            X_copy[col + "_isnull"] = X_copy[col].isnull().astype(int)
+        return X_copy
+
+
+class UserSpamClassifier(Resource):
+    def post(self):
+        data = request.get_json(force=True)
+
+        df = pd.DataFrame([data])
+
+        # numeric_features = [
+        #     "theme_id",
+        #     "color_scheme_id",
+        #     "projects_limit",
+        #     "namespace_id",
+        # ]
+        # categorical_features = [
+        #     "bot",
+        #     "last_sign_in_at",
+        #     "confirmed_at",
+        #     "last_activity_on",
+        #     "current_sign_in_at",
+        #     "can_create_group",
+        #     "can_create_project",
+        #     "two_factor_enabled",
+        #     "external",
+        #     "private_profile",
+        #     "is_admin",
+        #     "note",
+        # ]
+        text_features = [
+            "username",
+            "name",
+            "bio",
+            "location",
+            "website_url",
+            "public_email",
+            "organization",
+            "skype",
+            "linkedin",
+            "twitter",
+            "job_title",
+            "pronouns",
+            "work_information",
+            "email",
+            "commit_email",
+            "avatar_url",
+        ]
+
+        length_transformer = LengthTransformer(text_features)
+        df = length_transformer.transform(df)
+
+        is_null_transformer = IsNullTransformer(text_features)
+        df = is_null_transformer.transform(df)
+
+        df["combined_text"] = df[text_features].apply(
+            lambda row: " ".join(row.values.astype(str)), axis=1
+        )
+        text_features = ["combined_text"]
+
+        processed_data = preprocessing_pipeline.transform(df)
+
+        prediction = model.predict(processed_data)
+
+        prediction_label = 1 if prediction[0][0] > 0.5 else 0
+
+        return jsonify(
```

Comment:
This is probably where I had the most difficulty figuring out how to accomplish this in the development process. What you described was the initial plan. However, I also wanted to allow the model and pipeline to be updated without updating the whole Spamphiban service, similar to how it is done by GitLab's Spamcheck service, which will occasionally pull the newest model from a given S3 bucket. 

The problem is that GitLab's Spamcheck service doesn't use a pipeline, because they send events from GitLab to Spamcheck in the format the model expects, something we cannot do with GitLab System Hooks. That means that we need the pipeline, which cannot be imported nicely, at the moment. The best way I could figure out how to package the pipeline and model nicely was to use a container, unfortunately

It is less than ideal, but I don't want to hard-code the model and pipeline into Spamphibian for various reasons, and I don't want to provide a solution where code is arbitrarily imported and executed.